###############
#### SETUP ####
###############

# Defailt configfile. Can be changed at command-line (--configfile)
configfile: 'snakemake_config.yaml'




##################
#### PIPELINE ####
##################

MAC = [250]
PHENOTYPES = ['ampicillin','ciprofloxacin', 'trimethoprim.sulfamethoxazole']



rule all:
   input:
        '../vcf_files/annotated_output.vcf.gz',
        '../tree/alignment/annotated_output_biallelic_goodcontigs_remoutliers_synonymous_MAC10_refedit_subsample.vcf',
        '../figs/genomic_PCA.png',
        '../figs/SFS_combined_remoutliers.png',
        '../tree/alignment/annotated_output_biallelic_goodcontigs_remoutliers_synonymous_MAC10_refedit_subsample.min4.phy',
        '../figs/genome_tree_synonymous_MAC10.png',
        #expand("geno_pred/output/gwas_MAC{mac}_{pheno}_bslmm_probit.param.txt", mac=MAC, pheno = PHENOTYPES)



rule get_data:
    conda: 'envs/aws_cli.yaml'
    input:
    output:
        '../vcf_files/annotated_output.vcf.gz',
        '../vcf_files/filtered_output.vcf.gz',
        '../pangenome/pangenome_genomes_SRA_GCA.csv',
        '../pangenome/whole_pan_ecor_presence_absence.csv',
        '../pangenome/whole_pangenome.fasta',
        '../phenotype_matrix/phenotype_matrix_08302024.csv',
        '../phenotype_matrix/SRA_to_genome_name.csv',
        '../strains_metadata.csv',
        '../presence_absence_data/presence_absence_sfs.txt',
        '../presence_absence_data/presence_absence_all.bim',
        '../presence_absence_data/presence_absence_all.bed',
        '../presence_absence_data/presence_absence_all.fam',
        '../presence_absence_data/presence_absence_all.map'
    shell:
        'aws s3 cp s3://ecoli-pangenomics-data/Ecoli_AMR_GenotypePhenotype/ ../../Ecoli_AMR_GenotypePhenotype --recursive'

#workdir: "path/to/workdir"

# Snakefiles with rules for particular pieces of the pipeline
# See Snakefiles for details
include: 'rules/filtering.smk'
include: 'rules/popgen_analyses.smk'
include: 'rules/genomic_prediction.smk'
#include: '../geno_pred/Snakemake'
